File: Sources/SignalProtocol/Fingerprint.swift
import SignalFfi

public struct DisplayableFingerprint {
    public let formatted: String

    internal init(formatted: String) {
        self.formatted = formatted
    }
}

public struct ScannableFingerprint {
    public let encoding: [UInt8]

    internal init(encoding: [UInt8]) {
        self.encoding = encoding
    }

    public func compare(against other: ScannableFingerprint) throws -> Bool {
        var result: Bool = false
        try checkError(signal_fingerprint_compare(&result, encoding, encoding.count,
                                                  other.encoding, other.encoding.count))
        return result
    }
}

public struct Fingerprint {
    public let scannable: ScannableFingerprint
    public let displayable: DisplayableFingerprint

    internal init(displayable: DisplayableFingerprint, scannable: ScannableFingerprint) {
        self.displayable = displayable
        self.scannable = scannable
    }
}

public struct NumericFingerprintGenerator {
    private let iterations: Int

    public init(iterations: Int) {
        self.iterations = iterations
    }

    public func create(version: Int,
                       localIdentifier: [UInt8],
                       localKey: PublicKey,
                       remoteIdentifier: [UInt8],
                       remoteKey: PublicKey) throws -> Fingerprint {

        var obj: OpaquePointer?
        try checkError(signal_fingerprint_new(&obj, UInt32(iterations), UInt32(version),
                                              localIdentifier, localIdentifier.count,
                                              localKey.nativeHandle,
                                              remoteIdentifier, remoteIdentifier.count,
                                              remoteKey.nativeHandle))

        let fprintStr = try invokeFnReturningString {
            signal_fingerprint_display_string(obj, $0)
        }
        let displayable = DisplayableFingerprint(formatted: fprintStr)

        let scannableBits = try invokeFnReturningArray {
            signal_fingerprint_scannable_encoding(obj, $0, $1)
        }
        let scannable = ScannableFingerprint(encoding: scannableBits)
        try checkError(signal_fingerprint_destroy(obj))

        return Fingerprint(displayable: displayable, scannable: scannable)
    }
}

File: react-native/ts/Fingerprint.ts
//
// Copyright 2024 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import * as Native from './Native';
import { PublicKey } from './EcKeys';

function getNative(): any {
  return (globalThis as any).__libsignal_native;
}

/**
 * A displayable fingerprint (numeric string for human comparison).
 */
export class DisplayableFingerprint {
  private readonly displayString: string;

  constructor(displayString: string) {
    this.displayString = displayString;
  }

  /**
   * Get the displayable fingerprint string (e.g., "12345 67890 ...").
   */
  toString(): string {
    return this.displayString;
  }
}

/**
 * A scannable fingerprint (binary data for QR code comparison).
 */
export class ScannableFingerprint {
  private readonly _nativeHandle: Native.Fingerprint;

  constructor(nativeHandle: Native.Fingerprint) {
    this._nativeHandle = nativeHandle;
  }

  /**
   * Serialize this scannable fingerprint to bytes.
   */
  toBuffer(): Uint8Array {
    const n = getNative();
    return n.ScannableFingerprint_Serialize(this._nativeHandle);
  }

  /**
   * Compare this scannable fingerprint with another.
   * Returns true if they match.
   */
  compare(other: Uint8Array): boolean {
    const n = getNative();
    return n.ScannableFingerprint_Compare(
      this.toBuffer(),
      other
    );
  }
}

/**
 * A fingerprint for verifying identity keys.
 *
 * Contains both a displayable (numeric) representation and a scannable (binary)
 * representation for verifying that two parties have the same view of each
 * other's identity keys.
 */
export class Fingerprint {
  readonly _nativeHandle: Native.Fingerprint;

  private constructor(handle: Native.Fingerprint) {
    this._nativeHandle = handle;
  }

  /**
   * Create a new fingerprint.
   * @param iterations Number of hash iterations (typically 1024 or 5200)
   * @param version Protocol version
   * @param localIdentifier Local user's identifier (e.g., phone number hash)
   * @param localKey Local user's identity public key
   * @param remoteIdentifier Remote user's identifier
   * @param remoteKey Remote user's identity public key
   */
  static new(
    iterations: number,
    version: number,
    localIdentifier: Uint8Array,
    localKey: PublicKey,
    remoteIdentifier: Uint8Array,
    remoteKey: PublicKey
  ): Fingerprint {
    const n = getNative();
    return new Fingerprint(
      n.Fingerprint_New(
        iterations,
        version,
        localIdentifier,
        localKey._nativeHandle,
        remoteIdentifier,
        remoteKey._nativeHandle
      )
    );
  }

  /**
   * Get the displayable (numeric) fingerprint.
   */
  displayableFingerprint(): DisplayableFingerprint {
    const n = getNative();
    return new DisplayableFingerprint(
      n.Fingerprint_DisplayString(this._nativeHandle)
    );
  }

  /**
   * Get the scannable (binary) fingerprint.
   */
  scannableFingerprint(): ScannableFingerprint {
    return new ScannableFingerprint(this._nativeHandle);
  }
}
